jumanji/environments/routing/rware/generator.py

```diff
@@ -0,0 +1,227 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+from typing import Callable
+
+import chex
+import jax
+import jax.numpy as jnp
+
+from jumanji.environments.routing.rware import utils
+from jumanji.environments.routing.rware.types import State
+
+
+class Generator(abc.ABC):
+    """Base class for generators for the Rware environment."""
+
+    def __init__(
+        self,
+        shelf_rows: int,
+        shelf_columns: int,
+        column_height: int,
+        num_agents: int,
+        sensor_range: int,
+        request_queue_size: int,
+    ) -> None:
+        """Initialises an rware generator, used to generate grids for the Rware environment.
+
+        Args:
+            shelf_rows: the number of shelf cluster rows, each of height = column_height.
+                Defaults to 1.
+            shelf_columns: the number of shelf cluster columns, each of width = 2 cells
+                (must be an odd number). Defaults to 3.
+            column_height: the height of each shelf cluster. Defaults to 8.
+            num_agents: the number of agents (robots) operating on the warehouse floor.
+                Defaults to 2.
+            sensor_range: the receptive field around an agent          O O O
+                (e.g. 1 implies a 360 view of 1 cell around the    ->  O x O
+                agent's position cell)                                 O O O
+                Defaults to 1.
+            request_queue_size: the number of shelves requested at any
+                given time which remains fixed throughout environment steps. Defaults to 4.
+            time_limit: the maximum step limit allowed within the environment
+            renderer: an optional `RwareViewer` instance to render the environment, if left as
+                None a default viewer is created when render is called. Defaults to 500.
+        """
+        if shelf_columns % 2 != 1:
+            raise ValueError(
+                "Environment argument: `shelf_columns`, must be an odd number."
+            )
+
+        self._shelf_rows = shelf_rows
+        self._shelf_columns = shelf_columns
+        self._column_height = column_height
+        self._num_agents = num_agents
+        self._sensor_range = sensor_range
+        self._request_queue_size = request_queue_size
+
+        self._grid_size = (
+            (column_height + 1) * shelf_rows + 2,
+            (2 + 1) * shelf_columns + 1,
+        )
+        self._agent_ids = jnp.arange(num_agents)
+        self._make_warehouse()
+
+    def _make_warehouse(self) -> None:
```

Comment:
This seems like it shouldn't be in the abstract class, although maybe the warehouse should always be the same. I'm not sure if we'd ever want to generate different warehouse patterns

Reply:
Made a base class that has `_make_warehouse`, other can then choose to override the base class if they wish.